runtime/lua/vim/lsp/util.lua

```diff
@@ -1304,27 +1324,64 @@ function M.open_floating_preview(contents, syntax, opts)
     opts = { opts, 't', true };
   }
   opts = opts or {}
+  opts.fancy_markdown = opts.fancy_markdown == nil and true or opts.fancy_markdown
+  opts.close_events = opts.close_events or {"CursorMoved", "CursorMovedI", "BufHidden", "InsertCharPre"}
+
+  local bufnr = api.nvim_get_current_buf()
+
+  -- check if this popup is focusable and we need to focus
+  if opts.focus_id then
+    -- Go back to previous window if we are in a focusable one
+    local current_winnr = api.nvim_get_current_win()
+    if npcall(api.nvim_win_get_var, current_winnr, opts.focus_id) then
+      api.nvim_command("wincmd p")
+      return bufnr, current_winnr
+    end
+    do
+      local win = find_window_by_var(opts.focus_id, bufnr)
+      if win and api.nvim_win_is_valid(win) and vim.fn.pumvisible() == 0 then
+        -- focus and return the existing buf, win
+        api.nvim_set_current_win(win)
+        api.nvim_command("stopinsert")
+        return api.nvim_win_get_buf(win), win
+      end
+    end
+  end
+
+  local floating_bufnr = api.nvim_create_buf(false, true)
+  local is_fancy_markdown = syntax == "markdown" and opts.fancy_markdown
+
 
   -- Clean up input: trim empty lines from the end, pad
   contents = M._trim(contents, opts)
 
+  if is_fancy_markdown then
+    contents = M.format_fancy_markdown(floating_bufnr, contents, opts)
+  elseif syntax then
+    api.nvim_buf_set_option(floating_bufnr, 'syntax', syntax)
+  end
+
   -- Compute size of float needed to show (wrapped) lines
   opts.wrap_at = opts.wrap_at or (vim.wo["wrap"] and api.nvim_win_get_width(0))
   local width, height = M._make_floating_popup_size(contents, opts)
 
-  local floating_bufnr = api.nvim_create_buf(false, true)
-  if syntax then
-    api.nvim_buf_set_option(floating_bufnr, 'syntax', syntax)
-  end
   local float_option = M.make_floating_popup_options(width, height, opts)
   local floating_winnr = api.nvim_open_win(floating_bufnr, false, float_option)
-  if syntax == 'markdown' then
+  if syntax == 'markdown' or syntax == "lsp_markdown" then
```

Comment:
Because of this line, I'm saying we should only turn this on if stylize_markdown is true.

Reply:
I'm fine with changing that to something like `if syntax == "markdown" and opts.stylize_markdown then`.

Reply:
sgtm, if you make the changes we discussed re naming I think we're good after that.

Reply:
the focus_id thing notwithstanding, not sure what to do about that, it's probably fine in practice for now.